### PR TITLE
robot-name: add new exercise

### PR DESCRIPTION
**Pull request by filiptoskovic**
There is no canonical for this exercise. I'm not sure that tests are adequate for this exercise

### CODE REVIEW

**Comment by rpottsoh** [comment]
There has been some controversy surrounding this exercise.  Refer to the open and closed issues in repo problem-specifications.  As far as test data goes I would follow what csharp and/or ruby have done.  Since there is no test data there is nothing governing what each track comes up.  The wrinkle with this exercise is the requirement that no robots be named the same.  Somehow the robots need to know that when the come up with a name that no other robot is already using that name....  I have not seen anything in the tracks I have looked at, that have addressed this point.

I haven't looked at what you have done yet either.

File: exercises/robot-name/uRobotNameTest.pas [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  FRobot.DisposeOf;
+end;
+
+procedure TRobotNameTest.different_robots_have_different_names;
+var Names : TList<string>;
+  Robots : TObjectList<TRobot>;
+  i: Integer;
+begin
+  Names := TList<string>.create;
+  Robots := TObjectList<TRobot>.Create;
```
Comment (rpottsoh) [inline exercises/robot-name/uRobotNameTest.pas]: I think you need to invoke that `Robots` will take ownership of the objects that are being added.  That way the objects are properly destroyed when `Robots` is disposed of.  The way `TObjectList<>` is being used is no different than a `TList<>`

Comment (rpottsoh) [reply exercises/robot-name/uRobotNameTest.pas]: Never mind me!  I just looked up TObjectList and am reminded that by DEFAULT it assumes ownership of the objects added to it.  :+1:   Well played sir.

**Comment by rpottsoh** [comment]
Did you come up with this from scratch or did you model this on another track?  I like how you have gone about dealing with the dupe name issue.  I am not sure that the tests cover all the cases.  That is why I am asking if you modeled this implementation on another track's.  If not then I suggest looking in on Ruby, Csharp, any other tracks that may be implementing this exercise.

Nice job!

**Comment by rpottsoh** [comment]
I am going to move the call to Randomize down to the Initialization section.  Randomize should really only need to be called once and it is advisable to not call it inside of loops.

**Comment by rpottsoh** [comment]
I am happy with what I see.  Let me know when you are ready for this to be merged.  We can always add more tests later as we think of them.

**Comment by filiptoskovic** [comment]
I think that case is covered by `each_robot_have_unique_name` test. In that test 1000 robots is created. In case any robot name is not unique test will fail.

**Comment by filiptoskovic** [comment]
ok, uniqueness of name after reseting have to be tested.

**Comment by rpottsoh** [comment]
Would using a `TDictionary` simplify things.  The key could be the encoded integer and the value could be the robot name for the encoded integer.  It might simplify the process of looking up names to find repeats.  Is just a quick thought I haven't looked into it in any detail.

**Comment by rpottsoh** [comment]
I think what you have done is OK.  The exercise instructions state that the name must be generated randomly.  It doesn't state how to resolve repeat names.  Your way tries to minimize the time it takes to find an available name if a random name conflicts...  You could have easily just generated another random name, however it would be hard to say how may times it would take to come up with a non-conflicting name.  This exercise can always be revisited later if something comes up that we don't like.  If you are done I will merge.

**Comment by filiptoskovic** [comment]
Yes, it was my intention, but this is slow algorithm. This solution have to be rewritten. You can merge, but I will later change it. Thank you for spring4d. It have 4 patterns implemented. For now I have coded Strategy pattern example (not exercism example) and find patterns very useful. Actually I appears that I have invented hot water a lot of times.

**Comment by rpottsoh** [comment]
Spring4d is very nice.  I have implemented the observer pattern in my work along with the factory pattern.  The collections are much nicer as well, expanded above and beyond what is offered by Delphi.

**Pull request merged**